### PR TITLE
[dynamo] Implement `set.__contains__` for `Tensor` as object match of `FakeTensor`

### LINKED ISSUE #111556: [dynamo] Implement `set.__contains__` for tensors based on object identity
### 🚀 The feature, motivation and pitch

Workaround to https://github.com/pytorch/pytorch/issues/111544 for `set.__contains__` case

### Alternatives

_No response_

### Additional context

Related https://github.com/pytorch/pytorch/issues/111550

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305 @zou3519

### PR DESCRIPTION
**Pull request by jon-chuang**
Fixes https://github.com/pytorch/pytorch/issues/111556

Dynamo implementation of `set.__contains__` previously used `__eq__` match.

But this is wrong when `__eq__` match does not imply `__hash__` match, as is the case for `torch.Tensor`, leading to inconsistent results. See: https://github.com/pytorch/pytorch/issues/111542

Hence implement as Tensor object match i.e. proxy node `'example_value'` FakeTensor match.

This is valid since we guard on tensor aliasing.

cc

### CODE REVIEW

File: torch/_dynamo/variables/lists.py [hunk 1]
```diff
@@ -25,27 +25,47 @@
 from .functions import UserFunctionVariable, UserMethodVariable
 
 
-def _listlike_contains_helper(items, search, tx, options):
+def _listlike_contains_helper(items, search, tx, options, check_hash_match=False):
```
Comment (lezcano) [inline torch/_dynamo/variables/lists.py:28]: Why adding the `check_hash` as input, rather than always checking?

File: torch/_dynamo/variables/lists.py [hunk 2]
```diff
@@ -25,27 +25,47 @@
...
     if search.is_python_constant():
-        result = any(
+        found = any(
```
Comment (lezcano) [inline torch/_dynamo/variables/lists.py:30]: This path will probably fail if you are looking for a python constant on a listlike that just has tensors. Could you add a test?

File: torch/_dynamo/variables/lists.py [hunk 3]
```diff
@@ -883,7 +903,9 @@ def call_method(
         elif name == "__contains__":
             assert len(args) == 1
             assert not kwargs
-            return _listlike_contains_helper(self.items, args[0], tx, options)
+            return _listlike_contains_helper(
```
Comment (lezcano) [inline torch/_dynamo/variables/lists.py:906]: There are other calls to this function IIRC, do we need to change those as well?

File: torch/_dynamo/variables/lists.py (hunk 1 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py:28]: Because the behaviour for list is not to check object identity. 

Furthermore, we only `check_hash_match` for `Tensor` since we only know for a fact that tensor's hash is id and hence equivalent to its object identity.

We let the rest fail.

File: torch/_dynamo/variables/lists.py (hunk 2 above)
Comment (lezcano) [reply torch/_dynamo/variables/lists.py:30]: No, my point is that here you'll call `as_python_constant` on a tensor and it will raise an exception :D I think that you need to add an  `x.is_python_constant() and` in the condition

File: torch/_dynamo/variables/lists.py (hunk 1 above)
Comment (lezcano) [reply torch/_dynamo/variables/lists.py:28]: fair enough. All this will be properly fixed in https://github.com/pytorch/pytorch/pull/110524, but this LGTM as a workaround

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py:28]: We could extend this to `numpy.NdArray` if you would like @lezcano . Perhaps in follow-up PR.

File: test/dynamo/test_functions.py [hunk 4]
```diff
@@ -1835,6 +1835,24 @@ def fn(x):
...
+
+        tensor_list = set()
+        tensor_list.add(param2)
+        assert param not in tensor_list
+
+        def fn(param, param2):
+            tensor_list = set([param2])
+            return param in tensor_list
+
+        opt_fn = torch.compile(fn, fullgraph=True)
```
Comment (lezcano) [inline test/dynamo/test_functions.py:1850]: No need to fully torchcompile. Better compile it with `CompileCounter` and assert that there's one frame traced. I believe that `fullgraph=True` succeeds when there are no frames traced.

Otherwise, at least add `backend="eager"`

File: torch/_dynamo/variables/lists.py (hunk 2 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py:30]: Oh right. Makes sense.

File: torch/_dynamo/variables/lists.py [hunk 5]
```diff
@@ -25,27 +25,47 @@
...
+    must_check_hash = False
+    if check_hash_match and isinstance(search, variables.TensorVariable):
+        must_check_hash = True
+        # ID_MATCH of Tensor means ID_MATCH of proxies.
+
+        # There are scenarios in which this could potentially fail. That is when:
+        # 1. Previously unaliased is aliased e.g. f(x, y) -> f(x, x)
+        # 2. Previously aliased is unaliased e.g. f(x, x) -> f(y, y)
+        # Thankfully, we install guards which fail whenever the input tensors' alias matrix changes.
+        search = id(search.as_proxy())
```
Comment (jon-chuang) [inline torch/_dynamo/variables/lists.py:46]: @lezcano is it fine to id on the proxy itself, or is it better to id on the `proxy.node`, as is done in https://github.com/pytorch/pytorch/pull/111196?

Comment (lezcano) [reply torch/_dynamo/variables/lists.py:46]: I think you need to look at the node, as the proxy is created on the fly, while the node points to an honest-to-goodness node in the fx graph

File: test/dynamo/test_functions.py (hunk 4 above)
Comment (jon-chuang) [reply test/dynamo/test_functions.py:1850]: Hmm I think fullgraph=True is the same as nopython?

File: torch/_dynamo/variables/lists.py (hunk 2 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py:30]: Fixed in: https://github.com/pytorch/pytorch/pull/111572

File: torch/_dynamo/variables/lists.py (hunk 5 above)
Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: Actually, is this even good enough? I realize that if I apply mutations, things might fall apart:
```python
x = torch.zeros(4)
y = x.add_(z)
return x is y  # True in eager, False in compile
```

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: I guess we can graph break if the node is impure. Agreed? @lezcano

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: I think `example_value` which points to the `FakeTensor` is sufficient.

Comment (lezcano) [reply torch/_dynamo/variables/lists.py]: On a second thought, yes, that is right I think

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: I think the part where it doesn’t pass through vmap is actually a bug with fx and/or FakeTensor…

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: Turns out just passing `tensor.source` was good enough :laughing:

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: Turns out `tensor.source` and `node.meta['example_value']` are alternately the correct thing to do, but neither cover all cases :fearful:

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: I see, I think I figured it out.

When the tensor is created in the compiled function, then they will share the same source. Else, if the tensor is created outside the compiled region, then they will share the same FakeTensor :)

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: Ideally, they should always share the same source. I thin this is a bug. Let me try to fix :)

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: @lezcano so after some experiments in https://github.com/pytorch/pytorch/pull/111602, I've concluded that `TensorVariable.source` is probably the way to go. 

To fix it fully, we need to propagate it for all higher order ops. However, for this PR, we will just propagate for torch methods and functions.

File: torch/_dynamo/variables/torch.py [hunk 6]
```diff
@@ -236,6 +240,9 @@ def call_function(
         unspec_python_args = check_unspec_python_args(args, kwargs)
         options = VariableTracker.propagate(self, args, kwargs.values())
 
+        if self.value.__name__.endswith("_"):
```
Comment (jon-chuang) [inline torch/_dynamo/variables/torch.py:243]: Maybe you could split these out into separate PR?

Comment (jon-chuang) [reply torch/_dynamo/variables/torch.py:243]: btw please use the function instead

File: torch/_dynamo/variables/lists.py (hunk 5 above)
Comment (lezcano) [reply torch/_dynamo/variables/lists.py]: In my opinion, using the `example_value` would be better, as this one is generated directly by the meta calls of the relevant operators. This way if the meta operators are correct and the initial aliasing of locals and globals is correct, then everything will be inductively correct.

Comment (jon-chuang) [reply torch/_dynamo/variables/lists.py]: Yes, indeed, I made a mistake.

I think it's better if we fix the problem of higher order ops.

If we need a stronger guarantee, FakeTensor's storage can be a better signal for aliasing than FakeTensor object identity itself.

File: torch/_dynamo/variables/tensor.py [hunk 7]
```diff
@@ -348,10 +348,14 @@ def call_method(
                 unimplemented(f"Illegal method invocation {name} in strict mode")
         from . import ConstantVariable, TorchVariable, TupleVariable
         from .builder import wrap_fx_proxy
+        from .torch import is_torch_inplace_op
 
         kwargs = dict(kwargs)
         options = VariableTracker.propagate(self, args, kwargs.values())
 
+        if is_torch_inplace_op(name):  # inplace / self-propagating op
+            options.update({"source": self.source})
```
Comment (jon-chuang) [inline torch/_dynamo/variables/tensor.py:357]: This method is no longer viable. Please use `FakeTensor` identity instead. Fixing the higher-order ops issue is another matter.

**Comment by lezcano** [comment]
The bad rebase subscribed too many people. Please reopen a new PR.
